README
NAME
    Sub::Spec::To::Pod - Generate POD documentation from sub spec

VERSION
    version 0.22

SYNOPSIS
     % perl -MSub::Spec::To::Pod=gen_module_subs_pod \
         -e'print gen_module_subs_pod(module=>"MyModule")'

DESCRIPTION
    NOTICE: This module and the Sub::Spec standard is deprecated as of Jan
    2012. Rinci is the new specification to replace Sub::Spec, it is about
    95% compatible with Sub::Spec, but corrects a few issues and is more
    generic. Rias is the Perl implementation for Rinci and many of its
    modules can handle existing Sub::Spec sub specs. Rias::Sub::To::POD is
    the replacement for Sub::Spec::To::Pod.

    This module generates API POD documentation from sub specs in a
    specified module. Example specification:

     our %SPEC;

     $SPEC{sub1} = {
         summary     => 'Summary of sub1.',
         description => "Description of sub1 ...",
         args        => {
             arg1 => ['int*' => {
                 summary => 'Blah ...',
                 default => 0,
             }],
             arg2 => [str => {
                 summary => 'Blah blah ...',
                 ...
             }
         },
     }
     sub sub1 { ... }

     $SPEC{sub2} = { ... };
     sub sub2 { ... }

    Example output:

     =head2 sub1(%args) -> [STATUS_CODE, ERR_MSG, RESULT]

     Summary of sub1.

     Description of sub1...

     Arguments (* denotes required arguments):

     =over 4

     =item * arg1* => INT (default 0)

     Blah ...

     =item * arg2 => STR (default none)

     Blah blah ...

     =back

     =head2 sub2(%args) -> [STATUS_CODE, ERR_MSG, RESULT]

     ...

    This module uses Log::Any logging framework.

FUNCTIONS
    None of the functions are exported by default, but they are exportable.

  gen_module_subs_pod(%args) -> RESULT
    Arguments ("*" denotes required arguments):

    *   load => *bool* (default 1)

        Whether to load module using "require".

    *   module* => *str*

    *   path => *str*

        Specify exact path to load module (instead of relying on @INC).

    *   specs => *hash*

  spec_to_pod(%args) -> RESULT
    Generate POD documentation from sub spec.

    Arguments ("*" denotes required arguments):

    *   spec* => *hash*

SEE ALSO
    Sub::Spec

    Sub::Spec::To::HTML

    Sub::Spec::To::Org

    Sub::Spec::To::Text

AUTHOR
    Steven Haryanto <stevenharyanto@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Steven Haryanto.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Sub-Spec-To-Pod

0.22	 2012-01-31

	 No functional changes. Update to Data-Sah-0.02.


0.21	 2012-01-16

	 Cancel rename to Sub::Spec::To::Pod to avoid name clashes.


0.20	 2012-01-16

	 No functional changes. Add alias to old name Sub::Spec::To::Pod.


0.19     2012-01-16

         No functional changes. Rename from Sub::Spec::To::Pod to
         Sub::Spec::To::POD. Mark deprecated.


0.18     2011-11-23

         No functional changes. Replace Data::Sah::Util with Data::Sah.


0.17     2011-09-22

         Support 'args_as' spec clause: display arg_pos for subs with
         args_as =~ /array/.


0.16     2011-08-24

         Bug fixes (module location and function name).


0.15     2011-08-24

         Rename module to Sub::Spec::To::Pod, rename functions.


0.14     2011-04-07

         [ENHANCEMENTS]

         - Parse 'features' clause and note in the POD about these features:
           reverse, undo, pure, dry_run.


0.13     2011-03-08

         Fix small bug.

         Parse result_naked clause.


0.12     2011-02-23

         Add argument: specs.

         Rename argument: require -> load.


0.11     2011-02-22

         Use Sub::Spec::Utils


0.10     2011-02-18

         Generate documentation about response structure ([STATUSCODE, ERRMSG,
         RESULT] + a paragraph explaining it).

         Parse arg_aliases.


0.09     2011-02-11

         Remove mention of 'required_args'.


0.08     2011-02-11

         Build fix: missing dependency.


0.07     2011-02-10

         Formatting tweaks.


0.06     2011-02-10

         Packaging bug: backup files not removed.


0.05     2011-02-10

         Fix silly bug.


0.04     2011-02-10

         Fix schema clause ('in', not 'choices').


0.03     2011-02-10

         Formatting fixes & tweaks.


0.02     2011-02-10

         Add 'path' argument.

         Fix small typo.


0.01     2011-02-09

         First release.


lib/Sub/Spec/Pod.pm
package Sub::Spec::To::Pod;
BEGIN {
  $Sub::Spec::To::Pod::VERSION = '0.15';
}

use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use Sub::Spec::Utils; #tmp, for _parse_schema

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(spec_to_pod gen_module_subs_pod);

# VERSION

our %SPEC;

sub _parse_schema {
    Sub::Spec::Utils::_parse_schema(@_);
}

$SPEC{spec_to_pod} = {
    summary => 'Generate POD documentation from sub spec',
    args => {
        spec => ['hash*' => {}],
    },
    result_naked => 1,
};
sub spec_to_pod($;$) {
    # to minimize startup overhead
    require Data::Dump;
    require Data::Dump::Partial;
    require List::MoreUtils;

    my %args = @_;
    my $sub_spec = $args{spec} or return [400, "Please specify spec"];
    $log->trace("-> spec_to_pod($sub_spec->{_package}::$sub_spec->{name})");

    my $pod = "";

    die "No name in spec" unless $sub_spec->{name};
    $log->trace("Generating POD for $sub_spec->{name} ...");

    my $naked = $sub_spec->{result_naked};

    $pod .= "=head2 $sub_spec->{name}(\%args) -> ".
        ($naked ? "RESULT" : "[STATUS_CODE, ERR_MSG, RESULT]")."\n\n";

    if ($sub_spec->{summary}) {
        $pod .= "$sub_spec->{summary}.\n\n";
    }

    my $desc = $sub_spec->{description};
    if ($desc) {
        $desc =~ s/^\n+//; $desc =~ s/\n+$//;
        $pod .= "$desc\n\n";
    }

    if ($naked) {

    } else {
        $pod .= <<'_';
Returns a 3-element arrayref. STATUS_CODE is 200 on success, or an error code
between 3xx-5xx (just like in HTTP). ERR_MSG is a string containing error
message, RESULT is the actual result.

_
    }

    my $features = $sub_spec->{features} // {};
    if ($features->{reverse}) {
        $pod .= <<'_';
This function supports reverse operation. To reverse, add argument C<-reverse>
=> 1.

_
    }
    if ($features->{undo}) {
        $pod .= <<'_';
This function supports undo operation. See L<Sub::Spec::Clause::features> for
details on how to perform do/undo/redo.

_
    }
    if ($features->{dry_run}) {
        $pod .= <<'_';
This function supports dry-run (simulation) mode. To run in dry-run mode, add
argument C<-dry_run> => 1.

_
    }
    if ($features->{pure}) {
        $pod .= <<'_';
This function is declared as pure, meaning it does not change any external state
or have any side effects.

_
    }

    my $args  = $sub_spec->{args} // {};
    $args = { map {$_ => _parse_schema($args->{$_})} keys %$args };
    my $has_cat = grep { $_->{clause_sets}[0]{arg_category} }
        values %$args;

    if (scalar keys %$args) {
        my $noted_star_req;
        my $prev_cat;
        for my $name (sort {
            (($args->{$a}{clause_sets}[0]{arg_category} // "") cmp
                 ($args->{$b}{clause_sets}[0]{arg_category} // "")) ||
                     (($args->{$a}{clause_sets}[0]{arg_pos} // 9999) <=>
                          ($args->{$b}{clause_sets}[0]{arg_pos} // 9999)) ||
                              ($a cmp $b) } keys %$args) {
            my $arg = $args->{$name};
            my $ah0 = $arg->{clause_sets}[0];

            my $cat = $ah0->{arg_category} // "";
            if (!defined($prev_cat) || $prev_cat ne $cat) {
                $pod .= "=back\n\n" if defined($prev_cat);
                $pod .= ($cat ? ucfirst("$cat arguments") :
                             ($has_cat ? "General arguments":"Arguments"));
                $pod .= " (C<*> denotes required arguments)"
                    unless $noted_star_req++;
                $pod .= ":\n\n=over 4\n\n";
                $prev_cat = $cat;
            }

            $pod .= "=item * B<$name>".($ah0->{required} ? "*" : "")." => ";
            my $type;
            if ($arg->{type} eq 'any') {
                my @schemas = map {_parse_schema($_)} @{$ah0->{of}};
                my @types   = map {$_->{type}} @schemas;
                @types      = sort List::MoreUtils::uniq(@types);
                $type       = join("|", @types);
            } else {
                $type       = $arg->{type};
            }
            $pod .= "I<$type>";
            $pod .= " (default ".
                (defined($ah0->{default}) ?
                     "C<".Data::Dump::Partial::dumpp($ah0->{default}).">"
                         : "none").
                             ")"
                               if defined($ah0->{default});
            $pod .= "\n\n";

            my $aliases = $ah0->{arg_aliases};
            if ($aliases && keys %$aliases) {
                $pod .= "Aliases: ";
                my $i = 0;
                for my $al (sort keys %$aliases) {
                    $pod .= ", " if $i++;
                    my $alinfo = $aliases->{$al};
                    $pod .= "B<$al>".
                        ($alinfo->{summary} ? " ($alinfo->{summary})" : "");
                }
                $pod .= "\n\n";
            }

            $pod .= "Value must be one of:\n\n".
                join("", map {" $_\n"} split /\n/,
                     Data::Dump::dump($ah0->{in}))."\n\n"
                           if defined($ah0->{in});

            #my $o = $ah0->{arg_pos};
            #my $g = $ah0->{arg_greedy};

            $pod .= "$ah0->{summary}.\n\n" if $ah0->{summary};

            my $desc = $ah0->{description};
            if ($desc) {
                $desc =~ s/^\n+//; $desc =~ s/\n+$//;
                # XXX format/rewrap
                $pod .= "$desc\n\n";
            }
        }
        $pod .= "=back\n\n";

    } else {

        $pod .= "No known arguments at this time.\n\n";

    }

    $log->trace("<- spec_to_usage()");
    $pod;
}

$SPEC{gen_module_subs_pod} = {
    summary => '',
    args => {
        module => ['str*' => {}],
        specs => ['hash' => {}],
        load => ['bool' => {
            summary => 'Whether to load module using "require"',
            default => 1,
        }],
        path => ['str' => {
            summary => 'Specify exact path to load module '.
                '(instead of relying on @INC)',
        }],
    },
    result_naked => 1,
};
sub gen_module_subs_pod {
    my %args = @_;
    my $module = $args{module} or return [400, "Please specify module"];
    my $specs  = $args{specs};

    # require module and get specs
    my $modulep = $args{path};
    if (!defined($modulep)) {
        $modulep = $module;
        $modulep =~ s!::!/!g; $modulep .= ".pm";
    }
    if (!$specs) {
        if ($args{load} // 1) {
            $log->trace("Attempting to load $modulep ...");
            eval { require $modulep };
            die $@ if $@;
        }
        no strict 'refs';
        $specs = \%{$module."::SPEC"};
        #$log->tracef("\%$module\::SPEC = %s", $specs);
        die "Can't find \%SPEC in package $module\n" unless $specs;
    }
    $log->tracef("Functions that have spec: %s", [keys %$specs]);
    for (keys %$specs) {
        $specs->{$_}{_package} //= $module;
        $specs->{$_}{name}     //= $_;
    }

    join("", map { spec_to_usage(spec=>$specs->{$_}) } sort keys %$specs);
}

1;
# ABSTRACT: Generate POD documentation from sub spec


=pod

=head1 NAME

Sub::Spec::To::Pod - Generate POD documentation from sub spec

=head1 VERSION

version 0.15

=head1 SYNOPSIS

 % perl -MSub::Spec::To::Pod=gen_module_subs_pod \
     -e'print gen_module_subs_pod(module=>"MyModule")'

=head1 DESCRIPTION

This module generates API POD documentation from sub specs in a specified
module. Example specification:

 our %SPEC;

 $SPEC{sub1} = {
     summary     => 'Summary of sub1.',
     description => "Description of sub1 ...",
     args        => {
         arg1 => ['int*' => {
             summary => 'Blah ...',
             default => 0,
         }],
         arg2 => [str => {
             summary => 'Blah blah ...',
             ...
         }
     },
 }
 sub sub1 { ... }

 $SPEC{sub2} = { ... };
 sub sub2 { ... }

Example output:

 =head2 sub1(%args) -> [STATUS_CODE, ERR_MSG, RESULT]

 Summary of sub1.

 Description of sub1...

 Arguments (* denotes required arguments):

 =over 4

 =item * arg1* => INT (default 0)

 Blah ...

 =item * arg2 => STR (default none)

 Blah blah ...

 =back

 =head2 sub2(%args) -> [STATUS_CODE, ERR_MSG, RESULT]

 ...

This module uses L<Log::Any> logging framework.

=head1 FUNCTIONS

None of the functions are exported by default, but they are exportable.

=head1 SEE ALSO

L<Sub::Spec>

L<Sub::Spec::To::HTML>

L<Sub::Spec::To::Org>

L<Sub::Spec::To::Text>

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2011 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



lib/Sub/Spec/To/POD.pm
package Sub::Spec::To::POD;
{
  $Sub::Spec::To::POD::VERSION = '0.20';
}

use 5.010;
use strict;
use warnings;
use Log::Any '$log';

use Data::Sah;
use Lingua::EN::Numbers::Ordinate;

require Exporter;
our @ISA       = qw(Exporter);
our @EXPORT_OK = qw(spec_to_pod gen_module_subs_pod);

# VERSION

our %SPEC;

sub _parse_schema {
    Data::Sah::normalize_schema($_[0]);
}

$SPEC{spec_to_pod} = {
    summary => 'Generate POD documentation from sub spec',
    args => {
        spec => ['hash*' => {}],
    },
    result_naked => 1,
};
sub spec_to_pod($;$) {
    # to minimize startup overhead
    require Data::Dump;
    require Data::Dump::Partial;
    require List::MoreUtils;
    require Sub::Spec::Util;

    my %args = @_;
    my $sub_spec = $args{spec} or return [400, "Please specify spec"];
    $log->trace("-> spec_to_pod($sub_spec->{_package}::$sub_spec->{name})");

    my $pod = "";

    die "No name in spec" unless $sub_spec->{name};
    $log->trace("Generating POD for $sub_spec->{name} ...");

    my $naked = $sub_spec->{result_naked};

    my $pres = Sub::Spec::Util::parse_args_as($sub_spec->{args_as}//"hash");
    my $args_var = $pres->{args_var};
    $pod .= "=head2 $sub_spec->{name}($args_var) -> ".
        ($naked ? "RESULT" : "[STATUS_CODE, ERR_MSG, RESULT]")."\n\n";

    if ($sub_spec->{summary}) {
        $pod .= "$sub_spec->{summary}.\n\n";
    }

    my $desc = $sub_spec->{description};
    if ($desc) {
        $desc =~ s/^\n+//; $desc =~ s/\n+$//;
        $pod .= "$desc\n\n";
    }

    if ($naked) {

    } else {
        $pod .= <<'_';
Returns a 3-element arrayref. STATUS_CODE is 200 on success, or an error code
between 3xx-5xx (just like in HTTP). ERR_MSG is a string containing error
message, RESULT is the actual result.

_
    }

    my $features = $sub_spec->{features} // {};
    if ($features->{reverse}) {
        $pod .= <<'_';
This function supports reverse operation. To reverse, add argument C<-reverse>
=> 1.

_
    }
    if ($features->{undo}) {
        $pod .= <<'_';
This function supports undo operation. See L<Sub::Spec::Clause::features> for
details on how to perform do/undo/redo.

_
    }
    if ($features->{dry_run}) {
        $pod .= <<'_';
This function supports dry-run (simulation) mode. To run in dry-run mode, add
argument C<-dry_run> => 1.

_
    }
    if ($features->{pure}) {
        $pod .= <<'_';
This function is declared as pure, meaning it does not change any external state
or have any side effects.

_
    }

    my $args  = $sub_spec->{args} // {};
    $args = { map {$_ => _parse_schema($args->{$_})} keys %$args };
    my $has_cat = grep { $_->{clause_sets}[0]{arg_category} }
        values %$args;

    if (scalar keys %$args) {
        my $noted_star_req;
        my $prev_cat;
        for my $name (sort {
            (($args->{$a}{clause_sets}[0]{arg_category} // "") cmp
                 ($args->{$b}{clause_sets}[0]{arg_category} // "")) ||
                     (($args->{$a}{clause_sets}[0]{arg_pos} // 9999) <=>
                          ($args->{$b}{clause_sets}[0]{arg_pos} // 9999)) ||
                              ($a cmp $b) } keys %$args) {
            my $arg = $args->{$name};
            my $ah0 = $arg->{clause_sets}[0];

            my $cat = $ah0->{arg_category} // "";
            if (!defined($prev_cat) || $prev_cat ne $cat) {
                $pod .= "=back\n\n" if defined($prev_cat);
                $pod .= ($cat ? ucfirst("$cat arguments") :
                             ($has_cat ? "General arguments":"Arguments"));
                $pod .= " (C<*> denotes required arguments)"
                    unless $noted_star_req++;
                $pod .= ":\n\n=over 4\n\n";
                $prev_cat = $cat;
            }

            $pod .= "=item * B<$name>".($ah0->{req} ? "*" : "")." => ";
            my $type;
            if ($arg->{type} eq 'any') {
                my @schemas = map {_parse_schema($_)} @{$ah0->{of}};
                my @types   = map {$_->{type}} @schemas;
                @types      = sort List::MoreUtils::uniq(@types);
                $type       = join("|", @types);
            } else {
                $type       = $arg->{type};
            }
            $pod .= "I<$type>";
            $pod .= " (default ".
                (defined($ah0->{default}) ?
                     "C<".Data::Dump::Partial::dumpp($ah0->{default}).">"
                         : "none").
                             ")"
                               if defined($ah0->{default});
            $pod .= "\n\n";

            my $args_as = $sub_spec->{args_as} // 'hash';
            if ($args_as =~ /array/) {
                my $pos = $ah0->{arg_pos};
                my $greedy = $ah0->{arg_greedy};
                if (defined $pos) {
                    $pod .= ordinate($pos+1).
                        ($greedy ? " to the last argument(s)" : " argument");
                    if ($args_as =~ /ref/) {
                        if ($greedy) {
                            $pod .= " (\@{\$args}[$pos..last])";
                        } else {
                            $pod .= " (\$args->[$pos])";
                        }
                    } else {
                        if ($greedy) {
                            $pod .= " (\@args[$pos..last])";
                        } else {
                            $pod .= " (\$args[$pos])";
                        }
                    }
                    $pod .= ".\n\n";
                }
            }

            my $aliases = $ah0->{arg_aliases};
            if ($aliases && keys %$aliases) {
                $pod .= "Aliases: ";
                my $i = 0;
                for my $al (sort keys %$aliases) {
                    $pod .= ", " if $i++;
                    my $alinfo = $aliases->{$al};
                    $pod .= "B<$al>".
                        ($alinfo->{summary} ? " ($alinfo->{summary})" : "");
                }
                $pod .= ".\n\n";
            }

            $pod .= "Value must be one of:\n\n".
                join("", map {" $_\n"} split /\n/,
                     Data::Dump::dump($ah0->{in}))."\n\n"
                           if defined($ah0->{in});

            #my $o = $ah0->{arg_pos};
            #my $g = $ah0->{arg_greedy};

            $pod .= "$ah0->{summary}.\n\n" if $ah0->{summary};

            my $desc = $ah0->{description};
            if ($desc) {
                $desc =~ s/^\n+//; $desc =~ s/\n+$//;
                # XXX format/rewrap
                $pod .= "$desc\n\n";
            }
        }
        $pod .= "=back\n\n";

    } else {

        $pod .= "No known arguments at this time.\n\n";

    }

    $log->trace("<- spec_to_usage()");
    $pod;
}

$SPEC{gen_module_subs_pod} = {
    summary => '',
    args => {
        module => ['str*' => {}],
        specs => ['hash' => {}],
        load => ['bool' => {
            summary => 'Whether to load module using "require"',
            default => 1,
        }],
        path => ['str' => {
            summary => 'Specify exact path to load module '.
                '(instead of relying on @INC)',
        }],
    },
    result_naked => 1,
};
sub gen_module_subs_pod {
    my %args = @_;
    my $module = $args{module} or return [400, "Please specify module"];
    my $specs  = $args{specs};

    # require module and get specs
    my $modulep = $args{path};
    if (!defined($modulep)) {
        $modulep = $module;
        $modulep =~ s!::!/!g; $modulep .= ".pm";
    }
    if (!$specs) {
        if ($args{load} // 1) {
            $log->trace("Attempting to load $modulep ...");
            eval { require $modulep };
            die $@ if $@;
        }
        no strict 'refs';
        $specs = \%{$module."::SPEC"};
        #$log->tracef("\%$module\::SPEC = %s", $specs);
        die "Can't find \%SPEC in package $module\n" unless $specs;
    }
    $log->tracef("Functions that have spec: %s", [keys %$specs]);
    for (keys %$specs) {
        $specs->{$_}{_package} //= $module;
        $specs->{$_}{name}     //= $_;
    }

    join("", map { spec_to_pod(spec=>$specs->{$_}) } sort keys %$specs);
}

1;
# ABSTRACT: Generate POD documentation from sub spec


=pod

=head1 NAME

Sub::Spec::To::POD - Generate POD documentation from sub spec

=head1 VERSION

version 0.20

=head1 SYNOPSIS

 % perl -MSub::Spec::To::POD=gen_module_subs_pod \
     -e'print gen_module_subs_pod(module=>"MyModule")'

=head1 DESCRIPTION

B<NOTICE>: This module and the L<Sub::Spec> standard is deprecated as of Jan
2012. L<Rinci> is the new specification to replace Sub::Spec, it is about 95%
compatible with Sub::Spec, but corrects a few issues and is more generic.
L<Rias> is the Perl implementation for Rinci and many of its modules can handle
existing Sub::Spec sub specs. L<Rias::Sub::To::POD> is the replacement for
Sub::Spec::To::POD.

This module generates API POD documentation from sub specs in a specified
module. Example specification:

 our %SPEC;

 $SPEC{sub1} = {
     summary     => 'Summary of sub1.',
     description => "Description of sub1 ...",
     args        => {
         arg1 => ['int*' => {
             summary => 'Blah ...',
             default => 0,
         }],
         arg2 => [str => {
             summary => 'Blah blah ...',
             ...
         }
     },
 }
 sub sub1 { ... }

 $SPEC{sub2} = { ... };
 sub sub2 { ... }

Example output:

 =head2 sub1(%args) -> [STATUS_CODE, ERR_MSG, RESULT]

 Summary of sub1.

 Description of sub1...

 Arguments (* denotes required arguments):

 =over 4

 =item * arg1* => INT (default 0)

 Blah ...

 =item * arg2 => STR (default none)

 Blah blah ...

 =back

 =head2 sub2(%args) -> [STATUS_CODE, ERR_MSG, RESULT]

 ...

This module uses L<Log::Any> logging framework.

=head1 FUNCTIONS

None of the functions are exported by default, but they are exportable.

=head2 gen_module_subs_pod(%args) -> RESULT


Arguments (C<*> denotes required arguments):

=over 4

=item * B<load> => I<bool> (default C<1>)

Whether to load module using "require".

=item * B<module>* => I<str>

=item * B<path> => I<str>

Specify exact path to load module (instead of relying on @INC).

=item * B<specs> => I<hash>

=back

=head2 spec_to_pod(%args) -> RESULT


Generate POD documentation from sub spec.

Arguments (C<*> denotes required arguments):

=over 4

=item * B<spec>* => I<hash>

=back

=head1 SEE ALSO

L<Sub::Spec>

L<Sub::Spec::To::HTML>

L<Sub::Spec::To::Org>

L<Sub::Spec::To::Text>

=head1 AUTHOR

Steven Haryanto <stevenharyanto@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Steven Haryanto.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


__END__



Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.3601"
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Generate POD documentation from sub spec",
  "dist_author" => [
    "Steven Haryanto <stevenharyanto\@gmail.com>"
  ],
  "dist_name" => "Sub-Spec-To-Pod",
  "dist_version" => "0.22",
  "license" => "perl",
  "module_name" => "Sub::Spec::To::Pod",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Data::Dump::Partial" => 0,
    "Data::Sah" => "0.02",
    "JSON" => 0,
    "Lingua::EN::Numbers::Ordinate" => 0,
    "List::MoreUtils" => 0,
    "Log::Any" => 0,
    "Package::Alias" => 0,
    "Sub::Spec" => "1.0.4",
    "perl" => "5.010000"
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;
